Lib/zipfile.py

```diff
@@ -2228,11 +2229,18 @@ def joinpath(self, add):
 
     @staticmethod
     def _add_implied_dirs(names):
-        return names + [
+        subdirs = set([
             name + "/"
             for name in map(posixpath.dirname, names)
             if name and name + "/" not in names
-        ]
+        ])
+        missing_dirs = set()
+        for sd in list(subdirs):
+            for p in pathlib.PurePath(sd).parents:
+                if str(p) not in {".", "/"} and str(p) + "/" not in subdirs:
+                    missing_dirs.add(str(p) + "/")
+        subdirs.update(missing_dirs)
```

Comment:
Since subdirs is already a set, it's probably not necessary for missing_dirs to be a set (just an iterable).